aws-api-appsync/src/main/java/com/amplifyframework/datastore/appsync/ModelConverter.java

```diff
@@ -34,33 +34,64 @@ private ModelConverter() {}
 
     /**
      * Convert a Model to a Map&lt;String, Object&gt;.
-     * @param model a Model instance.
+     * @param instance a Model instance.
+     * @param schema a Model schema for the instance.
      * @param <T> type of the Model instance.
      * @return a Map&lt;String, Object&gt; representation of the provided Model instance.
+     * @throws AmplifyException if schema doesn't match instance
      */
-    public static <T extends Model> Map<String, Object> toMap(T model) {
-        if (model == null) {
-            return new HashMap<>();
+    public static <T extends Model> Map<String, Object> toMap(T instance, ModelSchema schema) throws AmplifyException {
+        final Map<String, Object> result = new HashMap<>();
+        for (ModelField modelField : schema.getFields().values()) {
+            String fieldName = modelField.getName();
+            try {
+                final ModelAssociation association = schema.getAssociations().get(fieldName);
+                if (association == null) {
+                    if (instance instanceof SerializedModel) {
+                        Map<String, Object> serializedData = ((SerializedModel) instance).getSerializedData();
+                        if (serializedData.containsKey(modelField.getName())) {
+                            result.put(fieldName, serializedData.get(modelField.getName()));
+                        }
+                    } else {
+                        result.put(fieldName, extractFieldValue(modelField, instance));
+                    }
+                } else if (association.isOwner()) {
+                    result.put(association.getTargetName(), extractAssociateId(modelField, instance));
+                }
+                // Ignore if field is associated, but is not a "belongsTo" relationship
+            } catch (Exception exception) {
+                throw new AmplifyException(
+                        "An invalid field was provided. " + fieldName + " is not present in " + schema.getName(),
+                        exception,
+                        "Check if this model schema is a correct representation of the fields in the provided Object");
+            }
         }
-        if (model instanceof SerializedModel) {
-            return ((SerializedModel) model).getSerializedData();
+        return result;
+    }
+
+    private static Object extractAssociateId(ModelField modelField, Model instance)
+            throws NoSuchFieldException, IllegalAccessException {
+        final Object fieldValue = extractFieldValue(modelField, instance);
+        if (modelField.isModel() && fieldValue instanceof Model) {
+            return ((Model) fieldValue).getId();
+        } else if (modelField.isModel() && fieldValue instanceof Map) {
```

Comment:
Is this a valid case to consider?

```java
} else if (modelField.isModel() && fieldValue instanceof SerializedModel) {
...
}
```

Reply:
No, at least not right now.  Right now, we don't ever construct trees with `SerializedModel`s.   A `SerializedModel` only stores the `id` of its children, not any of the child data.   I think we will need to do that in the future, to fully move from Java models to SerializedModel's, internally, but that's a bigger change.